### DISCUSSION TITLE
Setting Index on Declarative Base outside __table_args__

### DISCUSSION THREAD

**Discussion by carantunes** [sqlalchemy/sqlalchemy#7597] [Usage Questions]
**Describe the bug**
I came up against this behaviour trying to set up an Index for backend Postgres only, with DDL events, while omitting it for mysql, but it looks like no matter where I declare it, it's always mapped.

**Expected behaviour**
From the docs https://docs.sqlalchemy.org/en/13/core/constraints.html?highlight=index#setting-up-constraints-when-using-the-declarative-orm-extension I expected the Index to be mapped only if set inside `__table_args__.`

**To Reproduce**

```py
from sqlalchemy import Column, event, Boolean
from sqlalchemy import create_engine
from sqlalchemy import Index
from sqlalchemy import inspect
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.sql.ddl import CreateIndex

engine = create_engine("postgresql+psycopg2://invenio:invenio@localhost:5432/invenio", echo=True)
metadata = MetaData(engine)
Base = declarative_base(metadata=metadata)


def test_functional_indexes():
    Base = declarative_base(metadata=metadata)

    class Foo(Base):
        __tablename__ = 'foo'

        id = Column(Integer, primary_key=True)
        a = Column(String)
        b = Column(String)
        c = Column(Boolean)

        __table_args__ = (
            Index('ix_a', a),
        )

        ix_b = Index('ix_b', b, unique=True)

    ix_c = Index('ix_c', Foo.a, Foo.b, unique=True, postgresql_where=Foo.c)
    # Setting up DDL event throws Exception
    # sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateTable) relation "ix_c" already exists
    # event.listen(Foo.__table__, 'after_create', CreateIndex(ix_c).execute_if(dialect='postgresql'))

    Foo.__table__.drop(engine, checkfirst=True)
    Foo.__table__.create(engine)

    inspector = inspect(engine)
    indices = inspector.get_indexes(Foo.__tablename__)
    print(indices)  # <= Expected only "ix_a" to be in the list
    assert len(indices) == 1  # FAILS

```

**Versions.**
 - OS: MacOS
 - Python: Python 3.8.1
 - SQLAlchemy: 1.3.24
 - Database: Postgresql
 - DBAPI: psycopg2-binary
 - Alembic: 1.7.5

**Additional context**
After that, tried creating the index with DDL String to avoid the automap but then Alembic complains on alembic.compare_metadata().
```py

# DDL string used to avoid automap in mysql
create_ix_c = db.DDL("CREATE UNIQUE INDEX ix_c ON %(table)s (a, b) WHERE c")

db.event.listen(Foo.__table__, 'after_create', create_ix_c.execute_if(dialect=('postgresql'))

...
pprint.pprint(produce_migrations().upgrade_ops.as_diffs(), indent=2, width=20)

pprint.pprint(produce_migrations().downgrade_ops.as_diffs(), indent=2, width=20)

```
Output 
```
[ ( 'remove_index',
    Index('ix_c', Column('a', TEXT(), table=<foo>, nullable=False), Column('b', TEXT(), table=<foo>, nullable=False)))]

[ ( 'add_index',
    Index('ix_c', Column('a', TEXT(), table=<foo>, nullable=False), Column('b', TEXT(), table=<foo>, nullable=False)))]


```

**Have a nice day!**

**Comment by zzzeek** [comment]
indexes associate themselves with Table objects implicitly.  So if you want to make an Index that's oriented at "foo", but is not actually associated with that Table, you need to make a totally separate Table with a separate MetaData object like this:

```py
    ix_foo = Table("foo", MetaData(), Column("a"), Column("b"), Column("c"))
    ix_c = Index(
        "ix_c",
        ix_foo.c.a,
        ix_foo.c.b,
        unique=True,
        postgresql_where=ix_foo.c.c,
    )

    event.listen(
        Foo.__table__,
        "after_create",
        CreateIndex(ix_c).execute_if(dialect="postgresql"),
    )

```

this index wont be seen by alembic autogenerate, and in fact if you are using alembic migrations, there's no need to have this Index in your model metadata at all, just add it directly into the migration script in question.

**Reply by zzzeek** [reply]
as far as compare metadata,  to get that to work you would need to manually deal with this Index when you construct those metadata comparisons based on which type of database is in use.

basically the `__table_args__` thing is irrelevant here, those docs refer to "constraints" and in SQLAlchemy's terminiology "indexes" are not the same thing as "constraints".

**Comment by carantunes** [comment]
For anyone with the same issue In the end the only solution that worked was to keep the index creation as DDL string on an event and use alembic context's include_object to exclude said index entirely.

The feature I'd like was to simply omit index creation based on the backend in question, ie support returning null in the CreateIndex compiler, the same way CreateColumn supports skipping.

From the [DDL docs](https://docs.sqlalchemy.org/en/14/core/ddl.html#sqlalchemy.schema.CreateColumn):
> ... For example, suppose we wish to produce a Table which skips rendering of the PostgreSQL xmin column against the PostgreSQL backend, but on other backends does render it ...
```python
@compiles(CreateColumn, "postgresql")
def skip_xmin(element, compiler, **kw):
    if element.element.name == 'xmin':
        return None
    else:
        return compiler.visit_create_column(element, **kw)

```

Unfortunately compile on CreateIndex does not support return None and lead to `Query was empty` error.

**Reply by zzzeek** [reply]
the compiler approach is too low level and also too late in the process for this to be a good solution.  There is already a "conditional DDL" system that should work here.  There would just need to be some additional hint on the Index as to how it should be invoked in the DDL sequence.

to get alembic to interoperate better with "conditional DDL" elements, indexes and constraints should support being associated with the Table directly, so that Alembic can see them and eliminate the need for custom alembic include_object rules, but in such a way that it is known that they are part of a conditional system where they will not always be present.

it likely would mean that the rules we currently apply to [DDLElement.execute_if](https://docs.sqlalchemy.org/en/14/core/ddl.html?highlight=execute_if#sqlalchemy.schema.DDLElement.execute_if) can somehow be associated with SchemaItem objects directly.  that way both Alembic and SQLAlchemy's DDL invocation system can inspect them.

here's how that would look:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Index
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)

    __table_args__ = (
        Index("my_pg_index", data).ddl_if(dialect='postgresql'),
    )

e1 = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e1)


e2 = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e2)
Base.metadata.create_all(e2)

```

I will implement this for 2.0 as I already have a patch working.

**Reply by zzzeek** [reply]
see #7631 which has a proof of concept

**Reply by carantunes** [reply]
This is great! Exactly what I need, thank you!
